### PR TITLE
DR-780 Add flows and scripts for react app

### PR DESCRIPTION
**Pull request by timrickwood**
## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: infrastructure/stacks/test-stack-1/frontend.tf [hunk 1]
```diff
@@ -0,0 +1,39 @@
+module "front_end_bucket" {
```
Comment (jope1) [inline infrastructure/stacks/test-stack-1/frontend.tf:1]: General comment - I think I would have a separate test stack. 'ui-test-stack'?

File: .github/actions/derive-frontend-bucket/action.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+description: "Derives the name of the bucket used to store the SPA for subsequent actions to run against"
+
+inputs:
+    environment:
+        description: "The name of the environment to action the infrastructure"
+        required: true
+    workspace:
+        description: "The name of the workspace to action the infrastructure into."
+        required: true
+    repo_name:
```
Comment (jope1) [inline .github/actions/derive-frontend-bucket/action.yaml:11]: We usually refer to 'repo_name' as the 'domain'. (See deploy-infrastructure.yaml). Few occurrences of this in various actions and workflows

File: .github/workflows/react-infra-data.yaml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+          aws_account_id: ${{ secrets.ACCOUNT_ID }}
+          aws_region: ${{ vars.AWS_REGION }}
+
+      - name: Derive spa bucket
+        id: derive_spa_bucket
+        uses: NHSDigital/uec-dos-management/.github/actions/derive-frontend-bucket@v0.12
+        with:
+          repo_name: ${{ inputs.repo_name }}
+          environment: ${{ inputs.environment }}
+          workspace: ${{ inputs.workspace }}
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply .github/workflows/react-infra-data.yaml]: File removed and functionality moved to actions and the single reusable script as per review comments below

File: .github/workflows/react-infra-data.yaml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      spa_bucket_name:
+        description: "The name of the s3 bucket holding spa code"
+        value: "${{ jobs.derive-infra-data.outputs.spa_bucket_name }}"
+      cloudfront_dist_id:
+        description: "The id of the current cloudfront distribution"
+        value: "${{ jobs.derive-infra-data.outputs.cloudfront_dist_id }}"
+
+jobs:
+  derive-infra-data:
+    name: "Derive name of s3 bucket to host SPA"
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply .github/workflows/react-infra-data.yaml]: File removed and functionality moved to actions and the single reusable script as per review comments below

File: .github/workflows/react-infra-data.yaml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+      - name: Get CloudFront Distribution ID
+        id: get_cloudfront_dist_id
+        run: |
+          export ENVIRONMENT=${{ inputs.environment }}
+          export WORKSPACE=${{ inputs.workspace }}
+          export REPO_NAME=${{ inputs.repo_name }}
+          distribution_id=$(/bin/bash ./uec-dos-management/scripts/workflow/get-front-end-cloudfront-distribution.sh)
+          echo "distribution_id=${distribution_id}"
+          echo "distribution_id=${distribution_id}" >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply .github/workflows/react-infra-data.yaml]: Get dist id moved to new action and both steps moved to build and deploy react app yaml below

File: .github/workflows/build-and-deploy-react-app.yaml [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+      workspace:
+        description: "The name of the workspace to deploy the application into"
+        required: true
+        type: string
+      code_directory:
+        description: "The root directory for react code"
+        required: false
+        type: string
+        default: "src/frontend"
+      repo_name:
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply .github/workflows/build-and-deploy-react-app.yaml]: done

File: .github/workflows/build-and-deploy-react-app.yaml [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+      - name: Install dependencies
+        run: npm ci
+        working-directory: ${{ inputs.code_directory }}
+
+      - name: Run Linter
+        run: npm run lint
+        working-directory: ${{ inputs.code_directory }}
+
+      - name: Run Unit Tests
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply .github/workflows/build-and-deploy-react-app.yaml:86]: not addressed as part of this ticket

File: .github/workflows/build-and-deploy-react-app.yaml [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+          echo "Deploying to ${{ needs.react-infra-data.outputs.spa_bucket_name }}"
+          aws s3 sync build/ s3://${{ needs.react-infra-data.outputs.spa_bucket_name }}/
+        working-directory: ${{ inputs.code_directory }}
+
+      - name: Invalidate CloudFront Cache
+        shell: bash
+        run: |
+          echo ${{ needs.react-infra-data.outputs.cloudfront_dist_id }}
+          export DISTRIBUTION_ID=${{ needs.react-infra-data.outputs.cloudfront_dist_id }}
+          /bin/bash ./uec-dos-management/scripts/workflow/invalidate-cloudfront-cache.sh
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply .github/workflows/build-and-deploy-react-app.yaml:106]: Not addressed as part of this ticket

File: .github/actions/derive-frontend-bucket/action.yaml (hunk 2 above)
Comment (timrickwood) [reply .github/actions/derive-frontend-bucket/action.yaml]: repo_name changed to domain

File: infrastructure/stacks/test-stack-1/frontend.tf (hunk 1 above)
Comment (timrickwood) [reply infrastructure/stacks/test-stack-1/frontend.tf]: Separate test stack created - ui-test-stack

**Review by tamaragoldschmidt — APPROVED** [review]

**Review by jope1 — APPROVED** [review]

**Pull request merged**